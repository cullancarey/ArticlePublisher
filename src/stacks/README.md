# Python CDK Stack for Automated Article Publishing

This repository contains a Python CDK stack for automating the process of article publishing. The stack deploys an AWS Lambda function, IAM policies, and a CloudWatch Event Rule to publish articles at a scheduled time.

## Table of Contents

- [ArticlePublisherStack](#articlepublisherstack)

## ArticlePublisherStack

This stack sets up an automated workflow for publishing articles. The Lambda function is responsible for publishing articles and sharing them on social media platforms like LinkedIn and Twitter.

### Parameters

- `medium_api_token`: SSM parameter for the Medium API token.
- `medium_user_id`: SSM parameter for the Medium user ID.
- `openai_api_token`: SSM parameter for the OpenAI API token.
- `linkedin_access_token`: SSM parameter for the LinkedIn access token.
- `cullan_twitter_api_key`: SSM parameter for the Twitter API key.
- `cullan_twitter_access_token`: SSM parameter for the Twitter access token.
- `cullan_twitter_access_secret_token`: SSM parameter for the Twitter access secret token.
- `cullan_twitter_secret_key`: SSM parameter for the Twitter secret key.

### Features

- **Lambda Function**: Deploys a Dockerized Lambda function for article generation and sharing. This function runs on a schedule, and its logs are retained for one year.
  
- **IAM Policy**: Assigns an IAM policy to the Lambda function, allowing it to describe AWS pricing services.
  
- **SSM Parameter Access**: The Lambda function is granted read access to specified SSM parameters. This includes API tokens and other sensitive information for Medium, LinkedIn, OpenAI, and Twitter.

- **Scheduled Trigger**: Adds a CloudWatch Events Rule that triggers the Lambda function daily at 1:00 PM.