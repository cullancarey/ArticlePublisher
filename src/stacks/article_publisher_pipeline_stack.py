from aws_cdk import pipelines, Stack, Aws, Environment, App
from constructs import Construct
from stages.article_publisher_stage import ArticlePublisherPipelineStage


class ArticlePublisherPipelineStack(Stack):
    def __init__(self, scope: Construct, id: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        pipeline = pipelines.CodePipeline(
            self,
            "ArticlePublisherPipeline",
            synth=pipelines.ShellStep(
                "Synth",
                input=pipelines.CodePipelineSource.git_hub(
                    repo_string="cullancarey/ArticlePublisher", branch="develop"
                ),
                commands=[
                    "npm install -g aws-cdk",
                    "python -m pip install -r requirements.txt",
                    "cdk synth",
                ],
            ),
            docker_enabled_for_synth=True,
        )

        # Do this as many times as necessary with any account and region
        # Account and region may different from the pipeline's.
        pipeline.add_stage(ArticlePublisherPipelineStage(self, "Production"))
