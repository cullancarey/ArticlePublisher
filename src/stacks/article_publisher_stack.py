from aws_cdk import (
    Stack,
    aws_lambda as _lambda,
    aws_ssm as ssm,
    Duration,
    aws_logs as logs,
    aws_events as events,
    aws_events_targets as event_targets,
    aws_iam as iam,
)
from constructs import Construct


class ArticlePublisherStack(Stack):
    def __init__(
        self,
        scope: Construct,
        id: str,
        **kwargs,
    ) -> None:
        super().__init__(scope, id, **kwargs)

        # Lambda Function (Assuming that this is previously created)
        article_publisher_lambda = _lambda.DockerImageFunction(
            self,
            f"ArticlePublisherLambda",
            code=_lambda.DockerImageCode.from_image_asset(
                "assets/lambda/article_publisher/"
            ),
            timeout=Duration.seconds(120),
            architecture=_lambda.Architecture.X86_64,
            log_retention=logs.RetentionDays.ONE_YEAR,
        )

        # Define a policy statement
        statement = iam.PolicyStatement(
            actions=["pricing:DescribeServices"],
            resources=["*"],  # Adjust the resources as needed
            effect=iam.Effect.ALLOW,
        )

        # Add the policy statement to the Lambda function's execution role
        article_publisher_lambda.role.add_to_policy(statement)

        # Grant read access to the Lambda function for each SSM parameter

        _ = [
            ssm.StringParameter.from_secure_string_parameter_attributes(
                self, id=param, parameter_name=param
            ).grant_read(article_publisher_lambda)
            for param in [
                "medium_api_token",
                "medium_user_id",
                "openai_api_token",
                "linkedin_access_token",
                "cullan_twitter_api_key",
                "cullan_twitter_access_token",
                "cullan_twitter_secret_token",
                "cullan_twitter_secret_key",
            ]
        ]

        events.Rule(
            self,
            "ArticlePublisherLambdaRule",
            description=f"Rule to trigger {article_publisher_lambda.function_name}",
            schedule=events.Schedule.cron(
                week_day="*", hour="14", minute="0", month="*", year="*"
            ),
            targets=[event_targets.LambdaFunction(article_publisher_lambda)],
        ),
