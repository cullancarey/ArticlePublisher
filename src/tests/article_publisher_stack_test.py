from aws_cdk import App
from aws_cdk.assertions import Template

from stacks.article_publisher_stack import ArticlePublisherStack


def test_article_publisher_stack(snapshot):
    app = App()
    stack = ArticlePublisherStack(scope=app, id="TestArticlePublisherStack")

    template = Template.from_stack(stack)
    assert template.to_json() == snapshot
