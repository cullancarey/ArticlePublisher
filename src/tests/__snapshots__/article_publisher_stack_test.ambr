# serializer version: 1
# name: test_article_publisher_stack
  dict({
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'ArticlePublisherLambdaC6CEFD02': dict({
        'DependsOn': list([
          'ArticlePublisherLambdaServiceRoleDefaultPolicy370E1B9F',
          'ArticlePublisherLambdaServiceRoleAD42F84E',
        ]),
        'Properties': dict({
          'Architectures': list([
            'x86_64',
          ]),
          'Code': dict({
            'ImageUri': dict({
              'Fn::Sub': '${AWS::AccountId}.dkr.ecr.${AWS::Region}.${AWS::URLSuffix}/cdk-hnb659fds-container-assets-${AWS::AccountId}-${AWS::Region}:6470e5baebc5c5f2bdb3c3598a7146e495d75de4aab2d9e57a8888fdedcdafac',
            }),
          }),
          'PackageType': 'Image',
          'Role': dict({
            'Fn::GetAtt': list([
              'ArticlePublisherLambdaServiceRoleAD42F84E',
              'Arn',
            ]),
          }),
          'Timeout': 300,
        }),
        'Type': 'AWS::Lambda::Function',
      }),
      'ArticlePublisherLambdaLogRetention95816BF4': dict({
        'Properties': dict({
          'LogGroupName': dict({
            'Fn::Join': list([
              '',
              list([
                '/aws/lambda/',
                dict({
                  'Ref': 'ArticlePublisherLambdaC6CEFD02',
                }),
              ]),
            ]),
          }),
          'RetentionInDays': 365,
          'ServiceToken': dict({
            'Fn::GetAtt': list([
              'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A',
              'Arn',
            ]),
          }),
        }),
        'Type': 'Custom::LogRetention',
      }),
      'ArticlePublisherLambdaRuleAllowEventRuleTestArticlePublisherStackArticlePublisherLambda554020B7161270C4': dict({
        'Properties': dict({
          'Action': 'lambda:InvokeFunction',
          'FunctionName': dict({
            'Fn::GetAtt': list([
              'ArticlePublisherLambdaC6CEFD02',
              'Arn',
            ]),
          }),
          'Principal': 'events.amazonaws.com',
          'SourceArn': dict({
            'Fn::GetAtt': list([
              'ArticlePublisherLambdaRuleB5271BDC',
              'Arn',
            ]),
          }),
        }),
        'Type': 'AWS::Lambda::Permission',
      }),
      'ArticlePublisherLambdaRuleB5271BDC': dict({
        'Properties': dict({
          'Description': dict({
            'Fn::Join': list([
              '',
              list([
                'Rule to trigger ',
                dict({
                  'Ref': 'ArticlePublisherLambdaC6CEFD02',
                }),
              ]),
            ]),
          }),
          'ScheduleExpression': 'cron(0 13 ? * * *)',
          'State': 'ENABLED',
          'Targets': list([
            dict({
              'Arn': dict({
                'Fn::GetAtt': list([
                  'ArticlePublisherLambdaC6CEFD02',
                  'Arn',
                ]),
              }),
              'Id': 'Target0',
            }),
          ]),
        }),
        'Type': 'AWS::Events::Rule',
      }),
      'ArticlePublisherLambdaServiceRoleAD42F84E': dict({
        'Properties': dict({
          'AssumeRolePolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'sts:AssumeRole',
                'Effect': 'Allow',
                'Principal': dict({
                  'Service': 'lambda.amazonaws.com',
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'ManagedPolicyArns': list([
            dict({
              'Fn::Join': list([
                '',
                list([
                  'arn:',
                  dict({
                    'Ref': 'AWS::Partition',
                  }),
                  ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                ]),
              ]),
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Role',
      }),
      'ArticlePublisherLambdaServiceRoleDefaultPolicy370E1B9F': dict({
        'Properties': dict({
          'PolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'pricing:DescribeServices',
                'Effect': 'Allow',
                'Resource': '*',
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/medium_api_token',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/medium_user_id',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/openai_api_token',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/linkedin_access_token',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/cullan_twitter_api_key',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/cullan_twitter_access_token',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/cullan_twitter_access_secret_token',
                    ]),
                  ]),
                }),
              }),
              dict({
                'Action': list([
                  'ssm:DescribeParameters',
                  'ssm:GetParameters',
                  'ssm:GetParameter',
                  'ssm:GetParameterHistory',
                ]),
                'Effect': 'Allow',
                'Resource': dict({
                  'Fn::Join': list([
                    '',
                    list([
                      'arn:',
                      dict({
                        'Ref': 'AWS::Partition',
                      }),
                      ':ssm:',
                      dict({
                        'Ref': 'AWS::Region',
                      }),
                      ':',
                      dict({
                        'Ref': 'AWS::AccountId',
                      }),
                      ':parameter/cullan_twitter_secret_key',
                    ]),
                  ]),
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'PolicyName': 'ArticlePublisherLambdaServiceRoleDefaultPolicy370E1B9F',
          'Roles': list([
            dict({
              'Ref': 'ArticlePublisherLambdaServiceRoleAD42F84E',
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Policy',
      }),
      'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A': dict({
        'DependsOn': list([
          'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB',
          'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB',
        ]),
        'Properties': dict({
          'Code': dict({
            'S3Bucket': dict({
              'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}',
            }),
            'S3Key': '5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip',
          }),
          'Handler': 'index.handler',
          'Role': dict({
            'Fn::GetAtt': list([
              'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB',
              'Arn',
            ]),
          }),
          'Runtime': 'nodejs18.x',
          'Timeout': 900,
        }),
        'Type': 'AWS::Lambda::Function',
      }),
      'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB': dict({
        'Properties': dict({
          'AssumeRolePolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'sts:AssumeRole',
                'Effect': 'Allow',
                'Principal': dict({
                  'Service': 'lambda.amazonaws.com',
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'ManagedPolicyArns': list([
            dict({
              'Fn::Join': list([
                '',
                list([
                  'arn:',
                  dict({
                    'Ref': 'AWS::Partition',
                  }),
                  ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                ]),
              ]),
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Role',
      }),
      'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB': dict({
        'Properties': dict({
          'PolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': list([
                  'logs:PutRetentionPolicy',
                  'logs:DeleteRetentionPolicy',
                ]),
                'Effect': 'Allow',
                'Resource': '*',
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'PolicyName': 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB',
          'Roles': list([
            dict({
              'Ref': 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB',
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Policy',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
